### PR TITLE
Fix recursive unload, and fix hash calculation in load

### PR DESCRIPTION
**Pull request by callunity**
GP-3115
There were a couple issues at play here:
  * the calculation for finding which workflow runs take upstream workflow run files as input is corrected
  * the workflow version has was being calculated differently between `addWorkflowVersion` and `load`. I chose the `addWorkflowVersion` calculation as the correct one, partly because those hashes are already present in our databases.
    * I refactored to harmonize these calculations and a few other duplicated ones
  * When I initially generated test data for MainIntegrationTest, I generated fake hashes for workflow versions (and everything else). The `load` part of the unload-and-reload tests calculated the hashes that the workflow versions and workflow runs should have, so those hashes got updated when those tests were added. However, as per above, this calculation was incorrect, so all the test data hashes have been recalculated and corrected here.

### CODE REVIEW

File: vidarr-server/src/test/java/ca/on/oicr/gsi/vidarr/server/MainIntegrationTest.java [hunk 1]
```diff
@@ -1084,8 +1085,10 @@ public void whenCopyOut_thenRecordsAreCopied() {
             .then()
             .assertThat()
             .statusCode(200)
-            .body("workflowRuns.size()", is(8))
-            .body("workflowRuns.findAll { it.workflowName == \"bcl2fastq\" }.size()", is(8))
+            .body("workflowRuns.size()", greaterThanOrEqualTo(8))
```
[the remark this answers is not in the document]
Comment (callunity) [reply vidarr-server/src/test/java/ca/on/oicr/gsi/vidarr/server/MainIntegrationTest.java:1088]: Yes, if any new tests are added that use any of the existing bcl2fastq workflow runs as upstream data, or if any new bcl2fastq workflow runs are added for testing

File: vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/Main.java [hunk 2]
```diff
@@ -2587,6 +2571,7 @@ public Condition workflowRunId(
                 if (request.isRecursive()) {
                   Collection<Long> latestIds = workflowRuns;
                   do {
+                    var isDownstream = isDownstreamFrom(latestIds, configuration);
```
[the remark this answers is not in the document]
Comment (callunity) [reply vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/Main.java:2574]: Yes, that makes sense. I'll make that change!

[the remark this answers is not in the document]
Comment (callunity) [reply vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/Main.java:2574]: (though it'll be `downstreamIds = getWorkflowRunIdsDownstreamFron(latestIds,...)`, as the function starts with workflow runs and then searches for what is downstream/consumed by these workflow runs. I can change `latestIds` to `latestWorkflowRunIds` to clarify as well)

[the remark this answers is not in the document]
Comment (avarsava) [reply vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/Main.java:2574]: Ah okay, I'd parsed it as 'get the workflow ids from which latestIds is downstream', which would make all the ids returned upstream of latestIds. Thank you for the clarification!

File: vidarr-server/src/test/java/ca/on/oicr/gsi/vidarr/server/MainIntegrationTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (callunity) [reply vidarr-server/src/test/java/ca/on/oicr/gsi/vidarr/server/MainIntegrationTest.java:1088]: I'm going to switch it back to `equalTo` and update the numbers to reflect the newly-added test data

**Review by avarsava — APPROVED** [review]

File: vidarr-server/src/main/resources/db/migration/V0012__add_get_ids_for_downstream_workflow_runs_function.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
+-- The input file ids are Vidarr IDs formatted like: vidarr:<server>/file/<hash>
+-- Unnest the array of Vidarr IDs, then split each Vidarr ID on the slash and select the third item (<hash>)
+
+    CREATE OR REPLACE FUNCTION get_ids_for_downstream_workflow_runs(wr_ids BIGINT[]) RETURNS table (wfr_id bigint) AS $$
+      BEGIN
+        RETURN QUERY
+        SELECT DISTINCT id FROM workflow_run
+                WHERE array_length(input_file_ids, 1) > 0 AND
+                  -- workflow uses files produced by target workflow run as its input
+                  (SELECT ARRAY(SELECT DISTINCT SPLIT_PART(UNNEST(input_file_ids), '/', 3)::VARCHAR FROM workflow_run)) <@ (
```
Comment (mlaszloffy) [inline vidarr-server/src/main/resources/db/migration/V0012__add_get_ids_for_downstream_workflow_runs_function.sql:10]: Migration workflow runs currently do not use the `vidarr:<server>/file/<hash>` input id format. Either the ids need to be updated to use this format or coalescing on split_part and the unmodified hash id, something like:
```
select coalesce(nullif(vidarr_id,''), nullif(migration_id,'')) input_id, hash_id from (select split_part(unnest(input_file_ids),'/',3) vidarr_id, unnest(input_file_ids) migration_id, hash_id FROM workflow_run where cardinality(input_file_ids) > 0) as a;
```

Comment (mlaszloffy) [reply vidarr-server/src/main/resources/db/migration/V0012__add_get_ids_for_downstream_workflow_runs_function.sql:10]: Heather found that this was fixed and that migration in production is producing the expected input file pattern (vidarr:_/file/<input_file_hash_id>).

**Comment by callunity** [comment]
The last two commits are updating test data such that if this had been in place before the fix in " Fix the getIdsForDownstreamWorkflowRuns function in response to testi… 406e66a", the tests would have failed (correctly).

**Review by mlaszloffy — APPROVED** [review]

**Pull request merged**